### ISSUE TITLE
Recommended Parameters for > 3 million cells

### ISSUE THREAD

**Issue by MikhaelManurung** [KlugerLab/FIt-SNE#97]
Hello, I've been using the code and very impressed with how fast it is!

I am using FItSNE to embed millions of cells, usually around 3 - 9 million cells. I clustered my data with FlowSOM and then I used FItSNE to visualise my clusters. I have used the "standard" optimised settings, which is to scale the learning rate and perplexity with the number of cells as well as PCA initialization (in accordance with Kobak and Berens).

However, I noticed that my clusters are broken up.

![tsne](https://user-images.githubusercontent.com/10704760/79304909-a2fe9480-7ef2-11ea-87d2-2e93503986d6.png)

Below are the settings that I have used: 
```
tsne <- fftRtsne(
    as.matrix(X),
    max_iter = 1000,
    perplexity_list = list(30, nrow(exprs) / 100),
    initialization = pca_mat,
    learning_rate = nrow(exprs) / 12,
)
```

How should I proceed to optimise the settings? Is it because of insufficient early exaggeration iteration (I used the default 250)? Should I increase it to 2000 and then use 4000 max_iter?

In addition, should I also use late exaggeration?

```
start_late_exag_iter = 250
late_exag_coeff = 4 # or 2
mom_switch_iter = 250
```

I greatly appreciate your advice.

Best regards,
Mikhael

**Comment by dkobak** [comment]
Hi Mikhael. First of all, in the last FIt-SNE version (1.2.0) many of these parameters are default, so you if you update the version then you don't need to set them manually. Also, for such huge sample sizes there is little difference between perplexities 30 and 100, so I wouldn't bother with perplexity combination. I think you simply run `fftRtsne(as.matrix(X))` with the last version, you should get almost the same output as you have now (but a bit faster!).

I would definitely do a version with late exaggeration (`late_exag_coeff = 4`; all other parameters can be kept at default, `start_late_exag_iter = 250` is now default in version 1.2.0).

I would also suggest to select a random sample of 25000 points and do an embedding with large perplexity. In the paper we use `perplexity_list = list(30, 250)` so you can try that. Other parameters can be left default.

If I wanted to explore, I would look at these three embeddings.

If I wanted to produce only one single embedding without any exploration then I would use the approach we recommend in the paper: embed 25000 points as I wrote above, map all points on top of this embedding, use this as initialization for tSNE with exaggeration 4. (See https://github.com/berenslab/rna-seq-tsne/blob/master/demo.ipynb, last section, for the code in Python.)

**Comment by MikhaelManurung** [comment]
Hi Dmitry,

Thank you for the prompt response

**Comment by MikhaelManurung** [comment]
Hi Dmitry, thanks for the prompt response. The version of the provided Windows binary is still 1.1.0. Unfortunately, I do not have the skills to compile the build on my own. In addition, I can barely read Python code (I use R).

If you don't mind, and I would be really thankful for it, could you provide the parameters that I can use with FItSNE version 1.1.0?

**Comment by dkobak** [comment]
Oh I see. Yes there is some problem with Windows compilation of 1.2.0, I don't know why.

You have all the parameters correct in your initial post. But version 1.2.0 switched to using 750 iterations by default. So it would be `max_iter = 750, perplexity = 30, initialization = pca_mat, learning_rate = nrow(exprs) / 12` for the first run, which is more or less what you did already. Then use `start_late_exag_iter = 250, late_exag_coeff = 4` for the second run (the rest stays the same). For 25000-subsample version, use learning_rate 25000/12, perplexity combination of 30 and 250, PCA init.

Can you show the results when you get them? :-)

**Comment by MikhaelManurung** [comment]
Thank you! I will definitely post the results here as soon as possible.

**Comment by MikhaelManurung** [comment]
Hi Dmitry, here are the results!


![cyto_sce_myeloid_fitsne_parameterTuning](https://user-images.githubusercontent.com/10704760/79373399-cd823900-7f56-11ea-8742-64e0b94cb0fe.png)

In my opinion, late exaggeration does provide a more compact representation of my clusters. I am not sure what happened with the downsampled cells though...

**Comment by dkobak** [comment]
I cannot see any structure in the downsampled version that isn't visible in the full version. So I think in this case downsampling-based initialization is probably unnecessary. Whether you prefer to use late exaggeration or not, is up to you. You can also try some intermediate exaggeration values, if you like. The version with exaggeration=4 should be similar to what UMAP would give you.

Out of curiosity, what's the size of the data matrix here? Rows/columns?

**Comment by MikhaelManurung** [comment]
Thanks! I will try out other parameters as well. I really appreciate your inputs.

Would it be useful to do 2000 early iterations like in the [Nature Methods paper](https://www.nature.com/articles/s41592-018-0308-4/figures/8)?

For this data, I have ~4 million cells with 40 markers.

**Comment by dkobak** [comment]
> Would it be useful to do 2000 early iterations like in the Nature Methods paper?

No, I don't think so. They needed more early exaggeration iterations because they used the default learning rate. The more recent recommendation is to scale the learning rate with N (e.g. as N/early_exagg_coeff = N/12) instead of increasing the number of iterations. Similar end result, but much more efficient.
